File: Android/app/src/main/java/com/swarnkary/donezy/BackupCodec.kt
package com.swarnkary.donezy

import org.json.JSONArray
import org.json.JSONObject

/**
 * JSON envelope for full-database backup/restore. Photo URIs are kept as-is;
 * if the user restores onto a fresh device the file:// targets won't exist,
 * but text/rating data round-trips faithfully.
 */
object BackupCodec {

    private const val VERSION = 4

    fun encode(hobbies: List<Hobby>, logs: List<HobbyLog>): String {
        val root = JSONObject()
        root.put("version", VERSION)
        root.put("exportedAt", System.currentTimeMillis())

        val hobbiesArr = JSONArray()
        for (h in hobbies) {
            val (type, data) = h.recurrence.encode()
            hobbiesArr.put(JSONObject().apply {
                put("id", h.id)
                put("name", h.name)
                put("category", h.category)
                put("notes", h.notes)
                if (h.nextReminderAt != null) put("nextReminderAt", h.nextReminderAt)
                put("createdAt", h.createdAt)
                put("isPinned", h.isPinned)
                put("isArchived", h.isArchived)
                put("reminderIntervalHours", h.reminderIntervalHours)
                put("recurrenceType", type)
                put("recurrenceData", data)
                put("weeklyGoal", h.weeklyGoal)
            })
        }
        root.put("hobbies", hobbiesArr)

        val logsArr = JSONArray()
        for (l in logs) {
            logsArr.put(JSONObject().apply {
                put("id", l.id)
                put("hobbyId", l.hobbyId)
                put("entry", l.entry)
                put("createdAt", l.createdAt)
                if (l.rating != null) put("rating", l.rating)
                if (l.photoUri != null) put("photoUri", l.photoUri)
            })
        }
        root.put("logs", logsArr)

        return root.toString(2)
    }

    fun decode(json: String): Pair<List<Hobby>, List<HobbyLog>> {
        val root = JSONObject(json)
        val hobbiesArr = root.optJSONArray("hobbies") ?: JSONArray()
        val logsArr = root.optJSONArray("logs") ?: JSONArray()

        val hobbies = buildList {
            for (i in 0 until hobbiesArr.length()) {
                val o = hobbiesArr.getJSONObject(i)
                val type = o.optString("recurrenceType", "none").ifBlank { "none" }
                val data = o.optString("recurrenceData", "")
                val intervalHours = o.optLong("reminderIntervalHours", 0L)
                val recurrence = if (type == "none" && intervalHours > 0L) Recurrence.Hourly(intervalHours)
                                 else Recurrence.decode(type, data)
                add(
                    Hobby(
                        id = o.getLong("id"),
                        name = o.getString("name"),
                        category = o.optString("category", "General"),
                        notes = o.optString("notes", ""),
                        nextReminderAt = if (o.has("nextReminderAt") && !o.isNull("nextReminderAt"))
                            o.getLong("nextReminderAt") else null,
                        createdAt = o.optLong("createdAt", 0L),
                        isPinned = o.optBoolean("isPinned", false),
                        isArchived = o.optBoolean("isArchived", false),
                        reminderIntervalHours = intervalHours,
                        recurrence = recurrence,
                        weeklyGoal = o.optInt("weeklyGoal", 0)
                    )
                )
            }
        }
        val logs = buildList {
            for (i in 0 until logsArr.length()) {
                val o = logsArr.getJSONObject(i)
                add(
                    HobbyLog(
                        id = o.getLong("id"),
                        hobbyId = o.getLong("hobbyId"),
                        entry = o.optString("entry", ""),
                        createdAt = o.optLong("createdAt", 0L),
                        rating = if (o.has("rating") && !o.isNull("rating")) o.getInt("rating") else null,
                        photoUri = if (o.has("photoUri") && !o.isNull("photoUri")) o.getString("photoUri") else null
                    )
                )
            }
        }
        return hobbies to logs
    }
}

File: MacOs/Shared/BackupCodec.swift
import Foundation

/// JSON envelope for full-database backup/restore. Wire-compatible with the Android
/// `BackupCodec` (version 4) so backups round-trip between platforms. Photo URIs are
/// kept as-is; on a fresh device the file targets won't exist but text/rating data
/// round-trips faithfully.
enum BackupCodec {

    private static let version = 4

    static func encode(hobbies: [Hobby], logs: [HobbyLog]) -> String {
        var root: [String: Any] = [:]
        root["version"] = version
        root["exportedAt"] = nowMillis()

        root["hobbies"] = hobbies.map { h -> [String: Any] in
            let (type, data) = h.recurrence.encode()
            var o: [String: Any] = [
                "id": h.id,
                "name": h.name,
                "category": h.category,
                "notes": h.notes,
                "createdAt": h.createdAt,
                "isPinned": h.isPinned,
                "isArchived": h.isArchived,
                "reminderIntervalHours": h.reminderIntervalHours,
                "recurrenceType": type,
                "recurrenceData": data,
                "weeklyGoal": h.weeklyGoal,
            ]
            if let r = h.nextReminderAt { o["nextReminderAt"] = r }
            return o
        }

        root["logs"] = logs.map { l -> [String: Any] in
            var o: [String: Any] = [
                "id": l.id,
                "hobbyId": l.hobbyId,
                "entry": l.entry,
                "createdAt": l.createdAt,
            ]
            if let r = l.rating { o["rating"] = r }
            if let p = l.photoUri { o["photoUri"] = p }
            return o
        }

        let data = try? JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted, .sortedKeys])
        return data.flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
    }

    static func decode(_ json: String) throws -> (hobbies: [Hobby], logs: [HobbyLog]) {
        guard let data = json.data(using: .utf8),
              let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NSError(domain: "BackupCodec", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "Malformed backup file"])
        }

        let hobbiesArr = root["hobbies"] as? [[String: Any]] ?? []
        let logsArr = root["logs"] as? [[String: Any]] ?? []

        let hobbies: [Hobby] = hobbiesArr.compactMap { o in
            guard let id = (o["id"] as? NSNumber)?.int64Value,
                  let name = o["name"] as? String else { return nil }
            let type = (o["recurrenceType"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "none"
            let dataStr = o["recurrenceData"] as? String ?? ""
            let intervalHours = (o["reminderIntervalHours"] as? NSNumber)?.int64Value ?? 0
            let recurrence: Recurrence = (type == "none" && intervalHours > 0)
                ? .hourly(hours: intervalHours)
                : Recurrence.decode(type: type, data: dataStr)
            return Hobby(
                id: id,
                name: name,
                category: o["category"] as? String ?? "General",
                notes: o["notes"] as? String ?? "",
                nextReminderAt: (o["nextReminderAt"] as? NSNumber)?.int64Value,
                createdAt: (o["createdAt"] as? NSNumber)?.int64Value ?? 0,
                isPinned: o["isPinned"] as? Bool ?? false,
                isArchived: o["isArchived"] as? Bool ?? false,
                reminderIntervalHours: intervalHours,
                recurrence: recurrence,
                weeklyGoal: (o["weeklyGoal"] as? NSNumber)?.intValue ?? 0
            )
        }

        let logs: [HobbyLog] = logsArr.compactMap { o in
            guard let id = (o["id"] as? NSNumber)?.int64Value,
                  let hobbyId = (o["hobbyId"] as? NSNumber)?.int64Value else { return nil }
            return HobbyLog(
                id: id,
                hobbyId: hobbyId,
                entry: o["entry"] as? String ?? "",
                createdAt: (o["createdAt"] as? NSNumber)?.int64Value ?? 0,
                rating: (o["rating"] as? NSNumber)?.intValue,
                photoUri: o["photoUri"] as? String
            )
        }

        return (hobbies, logs)
    }
}
